### DISCUSSION TITLE
New default theme

### DISCUSSION THREAD

**Discussion by byrnereese** [mkdocs/mkdocs#3668] [Ideas]
I want to grow the mkdocs ecosystem of users and developers. I am an OG mkdocs user, but a watershed moment for me was the release of mkdocs-material. That product elevated the entire experience of using mkdocs, and output a website that I could easily and proudly share with coworkers. The quality of that product and the websites it produces rivals the best players in the industry.

New users to mkdocs, without having to adopt a third-party theme, should have a similar, high-quality, best-in-class experience.

I want to explore the idea of creating a new default mkdocs theme, or by blessing in some way a basic version of mkdocs-material as the official theme of mkdocs. We should not undermine third-parties and their ability to monetize the product (as @squidfunk has done so effectively). Quite the contrary - we need to make mkdocs a more attractive platform for people like @squidfunk to build a business or revenue stream on top of. 

One way we can do that is to make quality a central theme (no pun intended) of everything we do. 

(I should point out, that one facet of mkdocs-material that makes it such an amazing product is its documentation. As a tool that specializes in publishing documentation, our own documentation should model for the industry what good documentation looks like - just as mkdocs-material does.)

**Comment by pawamoy** [comment]
That is an excellent suggestion! We actually discussed about this previously with @squidfunk and @tomchristie. While this is a good idea, @squidfunk warned us that creating a theme is not hard, **but that making it endure on the long-term is extremely hard**. Users want *lots* of different things, you have to deal with different browsers and devices, etc. (@squidfunk is better positioned to speak about this).

It was suggested that maybe a mod of Material of MkDocs could be created to better fit MkDocs specifically, or more generally to allow to create a more unified style and experience across different projects of an organization. It was argued that (I'll shorten it) MM is already super modular, and simple but effective customizations could suffice.

About providing the same level of quality as MM for MkDocs itself, I definitely agree. While we don't think making MM *the* default theme would be a good idea (partly for the reasons you mention), I would myself advocate for adopting MM for MkDocs' own documentation :slightly_smiling_face: 

Another suggestion that was made during these discussions was to maybe split out the built-in `mkdocs` and `readthedocs` themes out of the core project. With good enough documentation, recommendations and user messages, we could have users choose a theme in order to use MkDocs. Something like this for themes managed within the @mkdocs org:

```
pip install mkdocs[classic]
pip install mkdocs[readthedocs]
```

...and the usual for third-party themes:

```
pip install mkdocs mkdocs-material
```

(introducing an extra like `pip install mkdocs[material]` would introduce a circular dependency, which are generally handled fine by tooling, but would maybe not be a good idea in terms of maintenance)

**Reply by squidfunk** [reply]
Author of Material for MkDocs here 👋 Thanks for all the praise, @byrnereese – I'm trying my best to live up to it ☺️

I absolutely agree – we should aim to make the out-of-the-box MkDocs experience better from the start. The current themes are dated, don't work well for large documentation projects and are just not that modular. I'm all for considering moving both built-in themes out of core into their own projects, because it would allow us to better focus on the internals of MkDocs, making it better for plugin and extension authors, and to decouple those themes from MkDocs release cycle. This would allow to more easily improve those themes, and MkDocs itself could adopt another theme for its own documentation. MkDocs is such an amazing ecosystem _because_ of its modularity and plugability – this is it's strength and competitive advantage compared to other SSGs. IMHO, we should focus on this, and __make mkdocs.org an entrypoint to its awesome ecosystem__.

Regarding a theme for MkDocs' own documentation: It does not have to be Material for MkDocs. However, it should be something modern that checks most boxes from the start. If we decide to use Material for MkDocs, a big upside is that maintenance on the entire frontend of the documentation is outsourced to me (and my team), so we could concentrate our efforts here on MkDocs itself. I would welcome this change, and would be happy to help with the migration on that part, as some reformatting might be necessary to use the best of Material for MkDocs.

If we decide on another theme, please understand that I'll likely not be able to help out that much, because building one theme already is my full-time job 😅 However, I'd be happy to give feedback how to achieve things, e.g. how certain things like dark mode are implemented in Material for MkDocs. I just want to warn everybody that aims to create a theme about the maintenance effort that it entails. If it wouldn't be for the funding, I would not be able to dedicate so much time of my life to Material for MkDocs, because keeping it a side project, it would fall apart quite quickly because of the sheer volume of requests and bug reports that are coming in.

Regarding installation of themes: if themes would make `mkdocs` a dependency, which it logically is, installation is simpler:

```
pip install mkdocs-classic
pip install mkdocs-readthedocs
pip install mkdocs-material
```

**Reply by Andre601** [reply]
> If we decide to use Material for MkDocs, a big upside is that maintenance on the entire frontend of the documentation is outsourced to me (and my team), so we could concentrate our efforts here on MkDocs itself.

The main downside is possible confusion. If the docs of MkDocs use the material theme, while the project itself by default does not will it cause confusion on the end-user, making them ask why the theme is so different...

So either Material would become the default theme for MkDocs itself, which may require a lot of extra work, or we work on a own, updated theme...

**Reply by squidfunk** [reply]
Tom, Timothee and I already talked about making mkdocs.org an entrypoint to the ecosystem. This would also mean restructuring its documentation to better explain what MkDocs is: an ecosystem. MkDocs could then use mkdocstrings to render its own API reference and Material for MkDocs as a theme – it's just a matter of explaining. This would however mean we need to move out the default themes, which has the other upsides I talked about in my comment.

**Reply by squidfunk** [reply]
On another note: we could also just start with Material for MkDocs, and if somebody cooks up a new theme that ticks all boxes (e.g. dark mode and friends), we could of course switch MkDocs own documentation to that in a second! We're currently trying find ways to move faster without having to invest months of work before we see an improvement. Nothing of this is however set in stone and I have absolutely no problem with MkDocs switching to or away from Material for MkDocs. I'm just thinking about economics.

**Reply by Andre601** [reply]
On the topic of redoing documentation, should we then utilize better alternative for existing extensions? Like instead of the build-in admonitions from Python-Markdown, utilize PyMdwonX's admonition blocks extensions?

Would allow a more readable raw markdown file for users to see...

**Reply by squidfunk** [reply]
Sure, if we redo the docs, we should discuss how we want to format them. However, changing markup that leads to the same result is probably not the most impactful direction that we can follow up on right now 😉

**Comment by Andre601** [comment]
Personally do I feel like the default theme (mkdocs) is fine as is for the most part.

There is only one part that I feel like should be looked into: Making a dark variant available for accessability.
The current theme is just... bright. This can hurt in the eyes over longer time (Especially if you suffer from light sensitivity with your eyes) of looking at it, and while tools such as Dark Reader seem to work quite well, is it resulting in the "true" colors of the theme being lost...

Also, with offering a dark theme, perhaps MkDocs could consider adding additional features for themes, that are usually only doable if the theme adds it themself such as a theme toggle and automatic toggle based on browser/OS preferences (Unless this is already available in MkDocs).

Another note I would like to give is, that with the [catalog](https://github.com/mkdocs/catalog) integration in MkDocs, installing themes, extensions, etc is more easy than ever... But the feature is not well promoted by MkDocs (Unless I, again, missed this somewhere), so maybe focus should be put on that too...

**Reply by pawamoy** [reply]
I am myself fine with keeping the `mkdocs` theme as default theme. However @squidfunk noted that if we use a different theme for MkDocs' own docs, users could be confused when they install it and see a different, less nice looking output. So yeah, that's a tough question. How to provide a good, default experience, that is the same as what MkDocs is using, without confusing users, and without impeding the work done by authors of other themes.

Regarding light/dark variants for the built-in themes, I think @waylan has been working on this. I worry that providing a general toggle feature from MkDocs itself could be quite hard: this depends a lot on the front-end stack used, which could be very different from one theme to another. We'll all agree though that if we can facilitate its implementation in third-party themes, that'd be great.

Great point about the catalog. We have plans for it: I've developed a [gallery of existing themes](https://github.com/pawamoy/mkdocs-gallery) that is to be [merged into the catalog](https://github.com/mkdocs/catalog/pull/165). Once it's merged, we could discuss integrating that more tightly within MkDocs own documentation, to provide much more visibility for existing themes :slightly_smiling_face:

**Reply by byrnereese** [reply]
> How to provide a good, default experience, that is the same as what MkDocs is using, without confusing users, and without impeding the work done by authors of other themes.

This to me in the central question. What I want to challenge the team to do is not to hold the status quo too dear. If we want to "take mkdocs to the next level" - then we need to take a very holistic approach, and understand what most people are in need of, and then align the base product to address 80% of those needs. 

I will admit though, I am not an expert in the default mkdocs theme. It was one of the first things I modified when I started using mkdocs simply because it didn't meet my needs. Of course, RingCentral's needs are perhaps atypical - in that the size and scale of our mkdocs documentation may be an outlier. My experience with it is therefore very superficial, and if I am being brutally honest, it doesn't have a significant "wow factor." Mkdocs is deserving of wow-inducing experience - end to end. 

When I think about the default theme, there are some basic things it should do really well:

* Responsiveness
* Accessibility
* Scalability

And it should look good. It might be an old stereotype, but software developers I believe tend to value function over form. But to grow mkdocs and to create a larger, more sustainable community on top of it, we need to recruit more companies like RingCentral to adopt it, and contribute to its advancement. To reach that audience, then aesthetics really do matter.

**Reply by byrnereese** [reply]
I was thinking more about what we want out of our default theme... it should embody the best practices of the product. Many developers will start simply on hacking on what comes with the product by default. Therefore, we want them hacking on a strong foundation, one that exhibits in clear and obvious ways many of the things mentioned above, including:

* **Modularity**. That means we develop and package various functional modules with the default theme. That way we provide a template for people to emulate and work from. 

Then as an activity, we can think about what the various modules might be, and what we even mean by "module." For example, the templates themselves are modular as they can be used to insert HTML and so forth deep in the guts of the theme. But then there are also plugins, which can include custom behaviors and so forth. 

One module may very well be: "light/dark mode toggle." I would be curious to hear from other what kind of features could be packaged in this way that might also represent new capabilities we want to add to the core product offering. 

Modules than can then become a very effective way for people to contribute new behaviors, and showcase what features are available through the mkdocs ecosystem.

**Reply by squidfunk** [reply]
This is a great idea I also already considered. However, the main problem is that templating (= frontend) always has to touch three domains: HTML, JS, and CSS. For example, a dark mode toggle would need markup, business logic and styling. We would effectively be building a component library, which is currently outside the scope of MkDocs, and which will likely not work with its current architecture. MkDocs only does pretty basic HTML rendering with the inclusion of static assets – that's it, and that is its strength. A component library would mean we would need to have a more complex build system like Docusaurus, which also means we would need to introduce a frontend build pipeline that would make building even slower. For instance, Docusaurus cannot handle the volume of pages that MkDocs does. If you talk to users and enterprise customers, for many of them, MkDocs is their very choice because it does not entail the necessity to have a JavaScript build system available – it's just good old Python ☺️

These matters are quite complex and lengthy to discuss. If somebody wants to talk about this, I'm happy to join a call where we can talk about the challenges that are related to coming down this path. Just let me know ☺️

**Reply by waylan** [reply]
> Regarding light/dark variants for the built-in themes, I think @waylan has been working on this.

Yes, I was. See mkdocs/mkdocs#3493. It got hung up on differences of opinion.

Regardless, I did work out a solution for light/dark toggles.  The difficulty is trying to maintain backward compatibility while introducing new features.

**Comment by fralau** [comment]
I believe this discussion is useful.

I am favorable for a new **basic** (beginner) theme to MkDocs. IMHO the purpose would not be to create a full-featured, aesthetic, enduring theme, since that role is taken by Material. That basic theme (as today) is a starter in the journey.

The primary requirement would to be super-simple (to lower the learning curve), while sufficiently pleasing to reward a user who uses MkDocs for the first time. And at some point, when they have outgrown that basic theme and they _need_ more advance features, they might want to move on to Material. 

Basically, I would most go for a revamp of the basic theme's css.

**Reply by fralau** [reply]
I would see this as a marketing factor: what is a complete beginner looking for when approaching MkDocs?

1. Something that can do the job.
2. Something that looks sufficiently professional.
3. Something they can understand.

And number 3 could make or break a new user. If they get the basic concept, they are on-board; if they are confused, they quit. That is why I believe we should really work on that "first impression", with its the (sometimes contradictory) requirements for a basic theme of **effectiveness, pleasantness and simplicity**.

Of the 3, simplicity ranks highest in my mind.

**Reply by byrnereese** [reply]
And I would add one more goal... the basic theme should also model best practices with regards to modularity and the proper use of templates. To the extent that we could even embed comments into the theme, that could also be helpful so that people can more easily customize it.

**Reply by pawamoy** [reply]
IMO you're describing Material for MkDocs :slightly_smiling_face: It does the job, it looks professional, it is well commented and can therefore be understood by most if not all. It's also modular and probably shows best-practices regarding theming and customization of themes.

**Reply by pawamoy** [reply]
The thing is, as a beginner on MkDocs I wouldn't want to spend time learning a basic theme to then throw away all my learnings and customizations when I realise I want to use Material for MkDocs instead.

Learning theming-related things is probably more interesting to people who want to *create* themes, or contribute to existing ones, not beginner users.

**Reply by pawamoy** [reply]
In any case, I am definitely not against have a better, simple default theme, which is not Material for MkDocs. Maybe the current default theme can simply be improved to fit the requirements you defined. Improvements have actually been done recently and will be merged very soon: https://github.com/mkdocs/mkdocs/pull/3493.

**Comment by byrnereese** [comment]
Is there a member of this team who is by training a designer? @squidfunk who did the design for material? Did you do that all yourself, or did you contract with a designer?

**Reply by squidfunk** [reply]
I did all off the designing of Material for MkDocs by myself, more or less following Google's Material Design guidelines (version 1). We're planning a rebrush/redesign later this year, taking the best parts of its current iteration (version 3).

**Reply by squidfunk** [reply]
Oh, and I need to say that I'm at best only a mediocre designer, and did not have any formal training 😅

**Reply by janbrasna** [reply]
Ooh, can't wait for the Material 3 redesign! That will be a killer update, is there an issue to track or an epic/board link? Thanks.
